src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala

```diff
@@ -98,28 +117,49 @@ object BlockTransactions extends ApiCodecs {
     for {
       headerId <- c.downField("headerId").as[ModifierId]
       transactions <- c.downField("transactions").as[List[ErgoTransaction]]
+      blockVersion <- c.downField("blockVersion").as[Version]
       size <- c.downField("size").as[Int]
-    } yield BlockTransactions(headerId, transactions, Some(size))
+    } yield BlockTransactions(headerId, blockVersion, transactions, Some(size))
   }
 }
 
 object BlockTransactionsSerializer extends ScorexSerializer[BlockTransactions] {
-
-  override def serialize(obj: BlockTransactions, w: Writer): Unit = {
-    w.putBytes(idToBytes(obj.headerId))
-    w.putUInt(obj.txs.size)
-    obj.txs.foreach { tx =>
+  // See a comment in the parse() function
+  val MagicNumber = 1000000
+
+  override def serialize(bt: BlockTransactions, w: Writer): Unit = {
+    w.putBytes(idToBytes(bt.headerId))
+    w.putUInt(MagicNumber + bt.blockVersion)
+    w.putUInt(bt.txs.size)
+    bt.txs.foreach { tx =>
       ErgoTransactionSerializer.serialize(tx, w)
     }
   }
 
   override def parse(r: Reader): BlockTransactions = {
     val startPos = r.position
     val headerId: ModifierId = bytesToId(r.getBytes(Constants.ModifierIdSize))
-    val size = r.getUInt().toIntExact
-    val txs = (1 to size).map { _ =>
+    val verOrCount = r.getUInt().toIntExact
+
+    /**
+      * A hack to avoid need for a database rescan if older version of the serializer was used to put.
+      * block transactions into.
+      *
+      * We consider that in the version 1 of the protocol there could not be a block with more
+      * than 1,000,000 transactions.
+      *
+      * Then the new serializer puts 1,000,000 + block version (while the old one just puts tx count with no version),
+      * and the reader knows that a new serializer was used if the first unsigned integer read is more than 1,000,000.
+      */
+    val (blockVersion, txCount) = if (verOrCount > MagicNumber) {
+      ((verOrCount - MagicNumber).toByte, r.getUInt().toIntExact)
+    } else {
+      (1: Byte, verOrCount)
+    }
+
+    val txs = (1 to txCount).map { _ =>
       ErgoTransactionSerializer.parse(r)
     }
```

Comment:
Please use cfor, it is even more natural here.

Reply:
Not related to this PR, please make issue to do this consistently around serializers